Static quantization for LSTM (front)

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32335 Benchmarks for quantized LSTM (static)
* **#32074 Static quantization for LSTM (front)**
* #31934 Static quantization for LSTM (back)

Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Static quantization for LSTM (front)

Commit:
Update on "[Draft] Static quantization for LSTM (front)"

Commit:
Update on "[WIP] Static quantization for LSTM (front)"

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 5568326c:


* **11/11** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=5568326c74daffa308adf31d2d25c9aebb3728f0).

### :detective: 9 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214137) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (1/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764091) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764091))
<details>
<summary>
<code>Jan 11 04:31:46 /pytorch/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member ‘kQ8BitZeroPoint’ declared ‘constexpr’</code>
</summary>

```
Jan 11 04:31:45        {weight_contig.size(0)}, weight_contig.options().dtype(at::kInt), LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 11 04:31:45                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 11 04:31:45 /pytorch/c10/core/MemoryFormat.h:34:36: note: declared here 
Jan 11 04:31:45  C10_DEPRECATED inline MemoryFormat get_contiguous_memory_format() { 
Jan 11 04:31:45                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 11 04:31:46 [ 53%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RangeFactories.cpp.o 
Jan 11 04:31:46 [ 53%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/ReduceOps.cpp.o 
Jan 11 04:31:46 /pytorch/aten/src/ATen/native/RNN.cpp:208:44: error: non-static data member ‘kQ8BitScaleWithRange16’ declared ‘constexpr’ 
Jan 11 04:31:46    constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:31:46                                             ^~~~~~ 
Jan 11 04:31:46 /pytorch/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member ‘kQ8BitZeroPoint’ declared ‘constexpr’ 
Jan 11 04:31:46    constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:31:46                                    ^ 
Jan 11 04:31:46 In file included from /pytorch/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 11 04:31:46                  from /pytorch/aten/src/ATen/Tensor.h:11, 
Jan 11 04:31:46                  from /pytorch/aten/src/ATen/Context.h:4, 
Jan 11 04:31:46                  from /pytorch/aten/src/ATen/ATen.h:5, 
Jan 11 04:31:46                  from /pytorch/aten/src/ATen/native/RNN.h:3, 
Jan 11 04:31:46                  from /pytorch/aten/src/ATen/native/RNN.cpp:1: 
Jan 11 04:31:46 /pytorch/aten/src/ATen/native/RNN.cpp: In function ‘std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor> at::native::_thnn_differentiable_lstm_cell_backward(const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&)’: 
Jan 11 04:31:46 /pytorch/c10/core/MemoryFormat.h:32:46: warning: ‘c10::MemoryFormat c10::get_contiguous_memory_format()’ is deprecated [-Wdeprecated-declarations] 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214136) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764092) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764092))
<details>
<summary>
<code>Jan 11 04:31:56 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr'</code>
</summary>

```
Jan 11 04:31:54 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qtanh.cpp.o 
Jan 11 04:31:54 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp.o 
Jan 11 04:31:54 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qclamp.cpp.o 
Jan 11 04:31:55 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconv.cpp.o 
Jan 11 04:31:55 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp.o 
Jan 11 04:31:56 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp.o 
Jan 11 04:31:56 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/fake_quantize_core.cpp.o 
Jan 11 04:31:56 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:208:44: error: non-static data member 'kQ8BitScaleWithRange16' declared 'constexpr' 
Jan 11 04:31:56    constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:31:56                                             ^ 
Jan 11 04:31:56 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr' 
Jan 11 04:31:56    constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:31:56                                    ^ 
Jan 11 04:31:56 In file included from /var/lib/jenkins/workspace/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 11 04:31:56                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Jan 11 04:31:56                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
Jan 11 04:31:56                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:5, 
Jan 11 04:31:56                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.h:3, 
Jan 11 04:31:56                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:1: 
Jan 11 04:31:56 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp: In function 'std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor> at::native::_thnn_differentiable_lstm_cell_backward(const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&)': 
Jan 11 04:31:56 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:32:46: warning: 'c10::MemoryFormat c10::get_contiguous_memory_format()' is deprecated [-Wdeprecated-declarations] 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214150) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (3/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764123) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764123))
<details>
<summary>
<code>Jan 11 04:31:52 /pytorch/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member ‘kQ8BitZeroPoint’ declared ‘constexpr’</code>
</summary>

```
Jan 11 04:31:50        {weight_contig.size(0)}, weight_contig.options().dtype(at::kInt), LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 11 04:31:50                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 11 04:31:50 /pytorch/c10/core/MemoryFormat.h:34:36: note: declared here 
Jan 11 04:31:50  C10_DEPRECATED inline MemoryFormat get_contiguous_memory_format() { 
Jan 11 04:31:50                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 11 04:31:51 [ 52%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RangeFactories.cpp.o 
Jan 11 04:31:52 [ 52%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/ReduceOps.cpp.o 
Jan 11 04:31:52 /pytorch/aten/src/ATen/native/RNN.cpp:208:44: error: non-static data member ‘kQ8BitScaleWithRange16’ declared ‘constexpr’ 
Jan 11 04:31:52    constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:31:52                                             ^~~~~~ 
Jan 11 04:31:52 /pytorch/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member ‘kQ8BitZeroPoint’ declared ‘constexpr’ 
Jan 11 04:31:52    constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:31:52                                    ^ 
Jan 11 04:31:52 [ 52%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/ReflectionPad.cpp.o 
Jan 11 04:31:52 In file included from /pytorch/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 11 04:31:52                  from /pytorch/aten/src/ATen/Tensor.h:11, 
Jan 11 04:31:52                  from /pytorch/aten/src/ATen/Context.h:4, 
Jan 11 04:31:52                  from /pytorch/aten/src/ATen/ATen.h:5, 
Jan 11 04:31:52                  from /pytorch/aten/src/ATen/native/RNN.h:3, 
Jan 11 04:31:52                  from /pytorch/aten/src/ATen/native/RNN.cpp:1: 
Jan 11 04:31:52 /pytorch/aten/src/ATen/native/RNN.cpp: In function ‘std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor> at::native::_thnn_differentiable_lstm_cell_backward(const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&)’: 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214131) pytorch_linux_xenial_py2_7_9_build (4/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764187) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764187))
<details>
<summary>
<code>Jan 11 04:32:31 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr'</code>
</summary>

```
Jan 11 04:32:29 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp.o 
Jan 11 04:32:29 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp.o 
Jan 11 04:32:29 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/fake_quantize_core.cpp.o 
Jan 11 04:32:30 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp.o 
Jan 11 04:32:30 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/init_qnnpack.cpp.o 
Jan 11 04:32:30 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconcat.cpp.o 
Jan 11 04:32:30 [ 55%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp.o 
Jan 11 04:32:31 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:208:44: error: non-static data member 'kQ8BitScaleWithRange16' declared 'constexpr' 
Jan 11 04:32:31    constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:32:31                                             ^~~~~~ 
Jan 11 04:32:31 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr' 
Jan 11 04:32:31    constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:32:31                                    ^ 
Jan 11 04:32:31 In file included from /var/lib/jenkins/workspace/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 11 04:32:31                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Jan 11 04:32:31                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
Jan 11 04:32:31                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:5, 
Jan 11 04:32:31                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.h:3, 
Jan 11 04:32:31                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:1: 
Jan 11 04:32:31 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp: In function 'std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor> at::native::_thnn_differentiable_lstm_cell_backward(const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&)': 
Jan 11 04:32:31 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:32:46: warning: 'c10::MemoryFormat c10::get_contiguous_memory_format()' is deprecated [-Wdeprecated-declarations] 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214141) pytorch_macos_10_13_py3_build (5/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764235) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764235))
<details>
<summary>
<code>Jan 10 20:32:38 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o</code>
</summary>

```
Jan 10 20:32:37   while (strides.size() < 2 * ntensors()) { 
Jan 10 20:32:37          ~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~ 
Jan 10 20:32:37 2 warnings generated. 
Jan 10 20:32:37 [1746/2989] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/TensorIteratorReduce.cpp.o 
Jan 10 20:32:37 ../aten/src/ATen/native/TensorIteratorReduce.cpp:137:23: warning: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Wsign-compare] 
Jan 10 20:32:37     for (int i = 0; i < non_reduced_shape.size(); ++i) { 
Jan 10 20:32:37                     ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~ 
Jan 10 20:32:37 1 warning generated. 
Jan 10 20:32:38 [1747/2989] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/TensorProperties.cpp.o 
Jan 10 20:32:38 [1748/2989] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o 
Jan 10 20:32:38 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o  
_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -c ../aten/src/ATen/native/RNN.cpp 
Jan 10 20:32:38 In file included from ../aten/src/ATen/native/RNN.cpp:6: 
Jan 10 20:32:38 In file included from ../aten/src/ATen/native/quantized/cpu/fbgemm_utils.h:5: 
Jan 10 20:32:38 ../third_party/fbgemm/include/fbgemm/QuantUtils.h:122:29: warning: comparison of integers of different signs: 'std::size_t' (aka 'unsigned long') and 'int' [-Wsign-compare] 
Jan 10 20:32:38   for (std::size_t i = 0; i < len; i++) { 
Jan 10 20:32:38                           ~ ^ ~~~ 
Jan 10 20:32:38 ../aten/src/ATen/native/RNN.cpp:208:3: error: non-static data member cannot be constexpr; did you intend to make it static? 
Jan 10 20:32:38   constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 10 20:32:38   ^ 
Jan 10 20:32:38   static  
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214143) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (6/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764840) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764840))
<details>
<summary>
<code>Jan 11 04:33:23 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o</code>
</summary>

```
Jan 11 04:33:22 ../c10/core/MemoryFormat.h:32:75: warning: ‘c10::MemoryFormat c10::get_contiguous_memory_format()’ is deprecated [-Wdeprecated-declarations] 
Jan 11 04:33:22  #define LEGACY_CONTIGUOUS_MEMORY_FORMAT c10::get_contiguous_memory_format() 
Jan 11 04:33:22                                                                            ^ 
Jan 11 04:33:22 ../aten/src/ATen/native/QuantizedLinear.cpp:251:73: note: in expansion of macro ‘LEGACY_CONTIGUOUS_MEMORY_FORMAT’ 
Jan 11 04:33:22        {weight_contig.size(0)}, weight_contig.options().dtype(at::kInt), LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 11 04:33:22                                                                          ^ 
Jan 11 04:33:22 ../c10/core/MemoryFormat.h:34:36: note: declared here 
Jan 11 04:33:22  C10_DEPRECATED inline MemoryFormat get_contiguous_memory_format() { 
Jan 11 04:33:22                                     ^ 
Jan 11 04:33:23 [1712/3016] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o 
Jan 11 04:33:23 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o  
X2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -c ../aten/src/ATen/native/RNN.cpp 
Jan 11 04:33:23 ../aten/src/ATen/native/RNN.cpp:208:44: error: non-static data member ‘kQ8BitScaleWithRange16’ declared ‘constexpr’ 
Jan 11 04:33:23    constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:33:23                                             ^ 
Jan 11 04:33:23 ../aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member ‘kQ8BitZeroPoint’ declared ‘constexpr’ 
Jan 11 04:33:23    constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:33:23                                    ^ 
Jan 11 04:33:23 In file included from aten/src/ATen/core/TensorBody.h:5:0, 
Jan 11 04:33:23                  from ../aten/src/ATen/Tensor.h:11, 
Jan 11 04:33:23                  from ../aten/src/ATen/Context.h:4, 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214147) pytorch_linux_xenial_py3_clang5_mobile_build (7/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764860) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764860))
<details>
<summary>
<code>Jan 11 04:33:32 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr'</code>
</summary>

```
Jan 11 04:33:31 [ 61%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/Copy.cpp.o 
Jan 11 04:33:31 [ 61%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/TensorFactories.cpp.o 
Jan 11 04:33:31 [ 62%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/QTensor.cpp.o 
Jan 11 04:33:32 [ 62%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qtanh.cpp.o 
Jan 11 04:33:32 [ 62%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconv_prepack.cpp.o 
Jan 11 04:33:32 [ 62%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qclamp.cpp.o 
Jan 11 04:33:32 [ 62%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconv.cpp.o 
Jan 11 04:33:32 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:208:44: error: non-static data member 'kQ8BitScaleWithRange16' declared 'constexpr' 
Jan 11 04:33:32    constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:33:32                                             ^ 
Jan 11 04:33:32 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr' 
Jan 11 04:33:32    constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:33:32                                    ^ 
Jan 11 04:33:32 In file included from /var/lib/jenkins/workspace/build_mobile/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 11 04:33:32                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Jan 11 04:33:32                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
Jan 11 04:33:32                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:5, 
Jan 11 04:33:32                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.h:3, 
Jan 11 04:33:32                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:1: 
Jan 11 04:33:32 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp: In function 'std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor> at::native::_thnn_differentiable_lstm_cell_backward(const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&)': 
Jan 11 04:33:32 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:32:46: warning: 'c10::MemoryFormat c10::get_contiguous_memory_format()' is deprecated [-Wdeprecated-declarations] 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214128) pytorch_linux_xenial_py3_6_gcc5_4_build (8/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764861) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764861))
<details>
<summary>
<code>Jan 11 04:33:30 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr'</code>
</summary>

```
Jan 11 04:33:29 /var/lib/jenkins/workspace/aten/src/ATen/native/LossMultiLabelMargin.cpp:338:42: note: in expansion of macro 'LEGACY_CONTIGUOUS_MEMORY_FORMAT' 
Jan 11 04:33:29    auto grad_input = at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 11 04:33:29                                           ^ 
Jan 11 04:33:29 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:34:36: note: declared here 
Jan 11 04:33:29  C10_DEPRECATED inline MemoryFormat get_contiguous_memory_format() { 
Jan 11 04:33:29                                     ^ 
Jan 11 04:33:29 [ 53%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/PointwiseOps.cpp.o 
Jan 11 04:33:30 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:208:44: error: non-static data member 'kQ8BitScaleWithRange16' declared 'constexpr' 
Jan 11 04:33:30    constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:33:30                                             ^ 
Jan 11 04:33:30 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:35: error: non-static data member 'kQ8BitZeroPoint' declared 'constexpr' 
Jan 11 04:33:30    constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:33:30                                    ^ 
Jan 11 04:33:30 In file included from /var/lib/jenkins/workspace/build/aten/src/ATen/core/TensorBody.h:5:0, 
Jan 11 04:33:30                  from /var/lib/jenkins/workspace/aten/src/ATen/Tensor.h:11, 
Jan 11 04:33:30                  from /var/lib/jenkins/workspace/aten/src/ATen/Context.h:4, 
Jan 11 04:33:30                  from /var/lib/jenkins/workspace/aten/src/ATen/ATen.h:5, 
Jan 11 04:33:30                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.h:3, 
Jan 11 04:33:30                  from /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:1: 
Jan 11 04:33:30 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp: In function 'std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor, at::Tensor> at::native::_thnn_differentiable_lstm_cell_backward(const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&, const at::Tensor&)': 
Jan 11 04:33:30 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:32:46: warning: 'c10::MemoryFormat c10::get_contiguous_memory_format()' is deprecated [-Wdeprecated-declarations] 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4214151) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (9/9)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83764904) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83764904))
<details>
<summary>
<code>Jan 11 04:33:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:3: error: non-static data member cannot be constexpr; did you intend to make it static?</code>
</summary>

```
Jan 11 04:33:41 3 warnings generated. 
Jan 11 04:33:41 [ 56%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconv.cpp.o 
Jan 11 04:33:42 [ 56%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_unpack.cpp.o 
Jan 11 04:33:42 [ 56%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qlinear_prepack.cpp.o 
Jan 11 04:33:42 [ 56%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/fake_quantize_core.cpp.o 
Jan 11 04:33:42 [ 56%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qconv_unpack.cpp.o 
Jan 11 04:33:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:208:3: error: non-static data member cannot be constexpr; did you intend to make it static? 
Jan 11 04:33:43   constexpr float kQ8BitScaleWithRange16 = 0.0625; 
Jan 11 04:33:43   ^ 
Jan 11 04:33:43   static  
Jan 11 04:33:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:209:3: error: non-static data member cannot be constexpr; did you intend to make it static? 
Jan 11 04:33:43   constexpr int kQ8BitZeroPoint = 0; 
Jan 11 04:33:43   ^ 
Jan 11 04:33:43   static  
Jan 11 04:33:43 /var/lib/jenkins/workspace/aten/src/ATen/native/RNN.cpp:1224:30: warning: 'get_contiguous_memory_format' is deprecated [-Wdeprecated-declarations] 
Jan 11 04:33:43     gog = at::zeros_like(cx, LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 11 04:33:43                              ^ 
Jan 11 04:33:43 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:32:46: note: expanded from macro 'LEGACY_CONTIGUOUS_MEMORY_FORMAT' 
Jan 11 04:33:43 #define LEGACY_CONTIGUOUS_MEMORY_FORMAT c10::get_contiguous_memory_format() 
Jan 11 04:33:43                                              ^ 
Jan 11 04:33:43 /var/lib/jenkins/workspace/c10/core/MemoryFormat.h:34:1: note: 'get_contiguous_memory_format' has been explicitly marked deprecated here 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[WIP] Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "[WIP] Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "[WIP] Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "[WIP] Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "[WIP] Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "[WIP] Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "[WIP] Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)

Commit:
Update on "Static quantization for LSTM (front)"


Differential Revision: [D19359209](https://our.internmc.facebook.com/intern/diff/D19359209)
